pallets/vesting/src/lib.rs

```diff
@@ -0,0 +1,291 @@
+			// 1646179200 = 2022-03-02 00:00:00(UTC)
+			(Percent::from_percent(11), 1646179200u64),
+			// 1651017600 = 2022-04-27 00:00:00(UTC)
+			(Percent::from_percent(11), 1651017600u64),
+			// 1655856000 = 2022-06-22 00:00:00(UTC)
+			(Percent::from_percent(11), 1655856000u64),
+			// 1660694400 = 2022-08-17 00:00:00(UTC)
+			(Percent::from_percent(11), 1660694400u64),
+		])
+		.unwrap_or_default()
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// The amount vested has been updated. This could indicate more funds are available.
+		/// The balance given is the amount which is left unvested (and thus locked).
+		/// \[account, unvested\]
+		VestingUpdated(T::AccountId, BalanceOf<T>),
+		/// An \[account\] has become fully vested. No further vesting can happen.
+		VestingCompleted(T::AccountId),
+		/// Update a vesting schedule.
+		/// \[new_schedule\]
+		VestingScheduleUpdated(BoundedVec<Schedule, T::MaxScheduleLength>),
+	}
+
+	/// Error for the vesting pallet.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The account given is not vesting.
+		NotVesting,
+		/// An existing vesting schedule already exists for this account that cannot be clobbered.
+		ExistingVestingSchedule,
+		/// Amount being transferred is too low to create a vesting schedule.
+		AmountLow,
+		/// Not enough tokens for vesting.
+		BalanceLow,
+		/// Cannot input
+		InvalidTimestamp,
+		/// The length of new schedule cannot be bigger/smaller than 7.
+		InvalidScheduleLength,
+		/// The new schedule should be sorted.
+		UnsortedSchedule,
+		/// The first round of vesting is not done yet.
+		ClaimTooEarly,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Update vesting schedule.
+		///
+		/// - `new_schedule`: New schedule for vesting.
+		#[pallet::weight(10_000)]
+		pub fn update_vesting_schedule(
+			origin: OriginFor<T>,
+			new_schedule: BoundedVec<Schedule, T::MaxScheduleLength>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			// We have only 7 rounds of schedule.
+			let old_schedule = VestingSchedule::<T>::get();
+			ensure!(
+				new_schedule.len() == old_schedule.len(),
+				Error::<T>::InvalidScheduleLength
+			);
+
+			// Ensure the new schedule should be sorted.
+			ensure!(
+				new_schedule.as_slice().windows(2).all(|w| w[0] < w[1]),
+				Error::<T>::UnsortedSchedule
+			);
+
+			let now = T::Timestamp::now().as_secs();
+			for (n, o) in new_schedule.iter().zip(old_schedule.iter()) {
+				// n == o means we will partialy update vesting schedule.
+				// n > o means new schedule is future schedule.
+				// n < o && n > now, also fine.
+				if *n == o.1 {
+					continue;
+				}
+				// This is an invalid schedule. New schedule cannot be past time.
+				if *n <= now {
+					return Err(Error::<T>::InvalidTimestamp.into());
+				}
+			}
```

Comment:
```suggestion
			for (n, o) in new_schedule.iter().zip(old_schedule.iter()) {
				if *n <= now {
				     ensure!(*n == o.1, ),
				     Error::<T>::InvalidTimestamp
				} 
			}
```

Reply:
You have to consider new schedule is future time. If we don't add `else` branch, it still can be possible to update a invalid schedule.
For example:
```
old-schedule = [1,2,3,4,5,6,7];
now = 3.5;
new-schedule = [1,2,3,2,3,2,2];
```
After check `[1,2,3]`, we still need to check the rest of new schedule.

```rust
for (n, o) in new_schedule.iter().zip(old_schedule.iter()) {
    if o.1 <= now {
	    ensure!(*n == o.1, Error::<T>::InvalidSchedule);
    } else {
	    ensure!(*n >= now, Error::<T>::InvalidSchedule);
    }
}
```

Reply:
You have to consider new schedule is future time. If we don't add `else` branch, it still can be possible to update a invalid schedule.
For example:
```
old-schedule = [10, 20, 30, 100, 110, 120, 130];
now = 99;
new-schedule = [10, 20, 30, 98, 110, 120, 130];
```
After check `[10, 20, 30]`, we still need to check the rest of new schedule.

```rust
for (n, o) in new_schedule.iter().zip(old_schedule.iter()) {
    if o.1 <= now {
        ensure!(*n == o.1, Error::<T>::InvalidSchedule);
    } else {
        ensure!(*n >= now, Error::<T>::InvalidSchedule);
    }
}
```